File: src/main/java/de/cplaiz/activecraftcore/guis/profilemenu/inventory/MainProfile.java
package de.cplaiz.activecraftcore.guis.profilemenu.inventory;

import de.cplaiz.activecraftcore.guicreator.GuiCreator;
import de.cplaiz.activecraftcore.guicreator.GuiCreatorDefaults;
import de.cplaiz.activecraftcore.guicreator.GuiItem;
import de.cplaiz.activecraftcore.guicreator.GuiNavigator;
import de.cplaiz.activecraftcore.guis.effectgui.EffectGui;
import de.cplaiz.activecraftcore.guis.profilemenu.ProfileMenu;
import de.cplaiz.activecraftcore.messagesv2.ColorScheme;
import de.cplaiz.activecraftcore.messagesv2.PlayerMessageFormatter;
import de.cplaiz.activecraftcore.messagesv2.MessageSupplier;
import de.cplaiz.activecraftcore.playermanagement.Profilev2;
import de.cplaiz.activecraftcore.utils.IntegerUtils;
import de.cplaiz.activecraftcore.utils.ItemBuilder;
import de.cplaiz.activecraftcore.utils.StringUtils;
import kotlin.Pair;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.ToString;
import org.bukkit.Material;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemFlag;
import org.bukkit.inventory.meta.PotionMeta;

import java.text.NumberFormat;

@Getter
@EqualsAndHashCode(callSuper = false)
@ToString
public class MainProfile extends GuiCreator {

    private static final String PREFIX = "profile.mainprofile.";
    private final ProfileMenu profileMenu;
    private final Player player, target;
    private final Profilev2 profile;
    private GuiItem connectionInfoStack, gameStats, violationStack, violationInfoStack, activeEffectsStack,
            gamemodeSwitcherStack, actionMenuStack, storageMenuStack, playerLocationStack, playtimeStack;
    private ItemBuilder activeEffectsBuilder;
    private final MessageSupplier messageSupplier;
    private final ColorScheme colorScheme;

    public MainProfile(ProfileMenu profileMenu) {
        super("main_profile", 6, profileMenu.getMessageSupplier().getMessage(PREFIX + "title"));
        this.profileMenu = profileMenu;
        this.player = profileMenu.getPlayer();
        this.target = profileMenu.getTarget();
        profile = profileMenu.getProfile();
        messageSupplier = profileMenu.getMessageSupplier();
        colorScheme = profileMenu.getColorScheme();
        refresh();
        profileMenu.setMainProfile(this);
    }

    @Override
    public void refresh() {
        // General Settings
        fillBackground(true);
        setItem(profileMenu.getDefaultGuiCloseItem(), 49);
        setItem(profileMenu.getPlayerHead(), 4);
        PlayerMessageFormatter msgFormatter = new PlayerMessageFormatter(GuiCreatorDefaults.acCoreMessage);
        int playtime = profile.getPlaytime();
        int playtimeMinutes = playtime % 60;
        int playtimeHours = (playtime - playtimeMinutes) / 60;
        msgFormatter
                .setTarget(profile)
                .addFormatterPatterns(new Pair<>("ip", target.getAddress().getHostName()), // TODO: 22.08.2022 testen ob das klappt. wenn nicht dann wieder "target.getAddress().getAddress().toString().replace("/", "")"
                        new Pair<>("port", target.getAddress().getPort() + ""),
                        new Pair<>("ping", target.getPing() + ""),
                        new Pair<>("health", target.getHealth() + ""),
                        new Pair<>("food", target.getFoodLevel() + ""),
                        new Pair<>("xp", target.getLevel() + ""),
                        new Pair<>("gamemode", StringUtils.toJadenCase(target.getGameMode().name())),
                        new Pair<>("bans", profile.getTimesBanned() + ""),
                        new Pair<>("ipbans", profile.getTimesIpBanned() + ""),
                        new Pair<>("warns", profile.getTimesWarned() + ""),
                        new Pair<>("mutes", profile.getTimesMuted() + ""),
                        new Pair<>("hours", playtimeHours + ""),
                        new Pair<>("minutes", playtimeMinutes + "")
                );

        // Player Items
        GuiItem slotEmpty = new GuiItem(Material.RED_STAINED_GLASS_PANE)
                .setDisplayName(messageSupplier.getMessage(PREFIX + "empty-slot"));
        GuiItem offHand = new GuiItem(target.getInventory().getItemInOffHand());
        GuiItem mainHand = new GuiItem(target.getInventory().getItemInMainHand());
        setItem(mainHand.getType() != Material.AIR ? mainHand : slotEmpty, 29);
        setItem(offHand.getType() != Material.AIR ? offHand : slotEmpty, 20);
        setItem(target.getInventory().getHelmet() != null ? new GuiItem(target.getInventory().getHelmet()) : slotEmpty, 10);
        setItem(target.getInventory().getChestplate() != null ? new GuiItem(target.getInventory().getChestplate()) : slotEmpty, 19);
        setItem(target.getInventory().getLeggings() != null ? new GuiItem(target.getInventory().getLeggings()) : slotEmpty, 28);
        setItem(target.getInventory().getBoots() != null ? new GuiItem(target.getInventory().getBoots()) : slotEmpty, 37);

        // Player Connection Information
        connectionInfoStack = new GuiItem(Material.STRUCTURE_VOID)
                .setDisplayName(messageSupplier.getMessage(PREFIX + "connection.title"))
                .setLore(messageSupplier.getFormatted(PREFIX + "connection.ip", msgFormatter),
                        messageSupplier.getFormatted(PREFIX + "connection.port", msgFormatter),
                        messageSupplier.getFormatted(PREFIX + "connection.ping", msgFormatter));
        setItem(connectionInfoStack, 21, player, "activecraft.connection.info");

        //player Stats
        NumberFormat formatter = NumberFormat.getInstance();
        formatter.setMaximumFractionDigits(2);
        gameStats = new GuiItem(Material.GRASS_BLOCK)
                .setDisplayName(messageSupplier.getMessage(PREFIX + "player.title"))
                .setLore(messageSupplier.getFormatted(PREFIX + "player.health", msgFormatter),
                        messageSupplier.getFormatted(PREFIX + "player.food", msgFormatter),
                        messageSupplier.getFormatted(PREFIX + "player.exp", msgFormatter),
                        messageSupplier.getFormatted(PREFIX + "player.gamemode", msgFormatter));
        setItem(gameStats, 12, player, "activecraft.stats.info");

        // Player Violations Information
        violationInfoStack = new GuiItem(Material.COMMAND_BLOCK)
                .setDisplayName(messageSupplier.getMessage(PREFIX + "violations.title"))
                .setLore(messageSupplier.getFormatted(PREFIX + "violations.bans", msgFormatter),
                        messageSupplier.getFormatted(PREFIX + "violations.ip-bans", msgFormatter),
                        messageSupplier.getFormatted(PREFIX + "violations.warns", msgFormatter),
                        messageSupplier.getFormatted(PREFIX + "violations.mutes", msgFormatter));
        setItem(violationInfoStack, 30, player, "activecraft.violations.info");

        // active effects
        activeEffectsBuilder = new ItemBuilder(Material.POTION)
                .displayname(messageSupplier.getMessage(PREFIX + "active-effects"));
        target.getActivePotionEffects().forEach(
                effect -> activeEffectsBuilder.lore(
                        messageSupplier.getMessage("effects." + effect.getType().getName().toLowerCase().replace("_", "-"))
                                + colorScheme.secondary() + "; "
                                + colorScheme.secondaryAccent() + effect.getAmplifier() + colorScheme.secondary() + "; "
                                + colorScheme.secondaryAccent() + IntegerUtils.shortInteger(effect.getDuration() / 20))
        );
        activeEffectsStack = new GuiItem(activeEffectsBuilder.build());
        PotionMeta potionMeta = (PotionMeta) activeEffectsStack.getItemMeta();
        potionMeta.addItemFlags(ItemFlag.HIDE_POTION_EFFECTS);
        activeEffectsStack.setItemMeta(potionMeta);
        // implement effectgui
        activeEffectsStack.addClickListener(guiClickEvent -> GuiNavigator.push(player, new EffectGui(player, target).getPotionEffectGui().build()));
        setItem(activeEffectsStack, 14, player, "activecraft.activeeffects");

        // Player Location Information
        // TODO: 22.08.2022 updaten
        playerLocationStack = new GuiItem(Material.REDSTONE_TORCH)
                .setDisplayName(messageSupplier.getMessage(PREFIX + "player-location"))
                .setLore(
                        colorScheme.secondaryAccent() + target.getWorld().getName() + colorScheme.secondary() + "; "
                                + colorScheme.secondaryAccent() + target.getLocation().getBlockX() + colorScheme.secondary() + ", "
                                + colorScheme.secondaryAccent() + target.getLocation().getBlockY() + colorScheme.secondary() + ", "
                                + colorScheme.secondaryAccent() + target.getLocation().getBlockZ()
                );
        setItem(playerLocationStack, 16, player, "activecraft.location");

        //Player Playtime
        playtimeStack = new GuiItem(Material.CLOCK)
                .setDisplayName(messageSupplier.getMessage(PREFIX + "playtime"))
                .setLore(messageSupplier.getFormatted(PREFIX + "playtime-lore", msgFormatter));
        setItem(playtimeStack, 15, player, "activecraft.playtime");


        // Other Menus
        // Player Punishment
        violationStack = new GuiItem(Material.COMMAND_BLOCK)
                .setDisplayName(messageSupplier.getMessage(PREFIX + "violations-gui"))
                .addClickListener(guiClickEvent -> GuiNavigator.push(player, profileMenu.getViolationsProfile().build()));
        setItem(violationStack, 42);

        // Player Gamemode Switcher
        gamemodeSwitcherStack = new GuiItem(Material.GRASS_BLOCK)
                .setDisplayName(messageSupplier.getMessage(PREFIX + "gamemode-switcher-gui"))
                .addClickListener(guiClickEvent -> GuiNavigator.push(player, profileMenu.getGamemodeSwitcherProfile().build()));
        setItem(gamemodeSwitcherStack, 32);

        // Player Storage Menu
        storageMenuStack = new GuiItem(Material.CHEST)
                .setDisplayName(messageSupplier.getMessage(PREFIX + "storage-gui"))
                .addClickListener(guiClickEvent -> GuiNavigator.push(player, profileMenu.getStorageProfile().build()));
        setItem(storageMenuStack, 33);

        // Player Actions Menu
        actionMenuStack = new GuiItem(Material.LEVER)
                .setDisplayName(messageSupplier.getMessage(PREFIX + "action-gui"))
                .addClickListener(guiClickEvent -> GuiNavigator.push(player, profileMenu.getActionProfile().build()));
        setItem(actionMenuStack, 34);
    }
}

File: src/main/java/org/activecraft/activecraftcore/guis/profilemenu/inventory/MainProfile.kt
package org.activecraft.activecraftcore.guis.profilemenu.inventory

import org.activecraft.activecraftcore.guicreator.GuiClickEvent
import org.activecraft.activecraftcore.guicreator.GuiCreator
import org.activecraft.activecraftcore.guicreator.GuiCreatorDefaults
import org.activecraft.activecraftcore.guicreator.GuiItem
import org.activecraft.activecraftcore.guicreator.GuiNavigator.Companion.push
import org.activecraft.activecraftcore.guis.effectgui.EffectGui
import org.activecraft.activecraftcore.guis.profilemenu.ProfileMenu
import org.activecraft.activecraftcore.messages.ColorScheme
import org.activecraft.activecraftcore.messages.MessageSupplier
import org.activecraft.activecraftcore.messages.PlayerMessageFormatter
import org.activecraft.activecraftcore.playermanagement.Profile
import org.activecraft.activecraftcore.utils.ItemBuilder
import org.activecraft.activecraftcore.utils.shortInteger
import org.activecraft.activecraftcore.utils.toJadenCase
import org.bukkit.Material
import org.bukkit.entity.Player
import org.bukkit.inventory.ItemFlag
import org.bukkit.inventory.meta.PotionMeta
import org.bukkit.potion.PotionEffect
import java.text.NumberFormat

class MainProfile(private val profileMenu: ProfileMenu) : GuiCreator(
    identifier = "main_profile",
    rows = 6,
    title = profileMenu.messageSupplier.getMessage(PREFIX + "title")
) {
    private val player: Player = profileMenu.player
    private val target: Player = profileMenu.target
    private val profile: Profile = profileMenu.profile
    private var connectionInfoStack: GuiItem? = null
    private var gameStats: GuiItem? = null
    private var violationStack: GuiItem? = null
    private var violationInfoStack: GuiItem? = null
    private var activeEffectsStack: GuiItem? = null
    private var gamemodeSwitcherStack: GuiItem? = null
    private var actionMenuStack: GuiItem? = null
    private var storageMenuStack: GuiItem? = null
    private var playerLocationStack: GuiItem? = null
    private var playtimeStack: GuiItem? = null
    private lateinit var activeEffectsBuilder: ItemBuilder
    private val messageSupplier: MessageSupplier = profileMenu.messageSupplier
    private val colorScheme: ColorScheme = profileMenu.colorScheme

    init {
        refresh()
    }

    override fun refresh() {
        // General Settings
        fillBackground(true)
        setItem(profileMenu.defaultGuiCloseItem, 49)
        setItem(profileMenu.playerHead, 4)
        val msgFormatter = PlayerMessageFormatter(GuiCreatorDefaults.acCoreMessage!!)
        val playtime = profile.playtime
        val playtimeMinutes = playtime % 60
        val playtimeHours = (playtime - playtimeMinutes) / 60
        msgFormatter
            .setTarget(profile)
            .addFormatterPatterns(
                Pair(
                    "ip",
                    target.address.hostName
                ),  // TODO: 22.08.2022 testen ob das klappt. wenn nicht dann wieder "target.getAddress().getAddress().toString().replace("/", "")"
                Pair("port", target.address.port.toString() + ""),
                Pair("ping", target.ping.toString() + ""),
                Pair("health", target.health.toString() + ""),
                Pair("food", target.foodLevel.toString() + ""),
                Pair("xp", target.level.toString() + ""),
                Pair("gamemode", toJadenCase(target.gameMode.name)),
                Pair("bans", profile.timesBanned.toString() + ""),
                Pair("ipbans", profile.timesIpBanned.toString() + ""),
                Pair("warns", profile.timesWarned.toString() + ""),
                Pair("mutes", profile.timesMuted.toString() + ""),
                Pair("hours", playtimeHours.toString() + ""),
                Pair("minutes", playtimeMinutes.toString() + "")
            )

        // Player Items
        val slotEmpty = GuiItem(Material.RED_STAINED_GLASS_PANE)
            .setDisplayName(messageSupplier.getMessage(PREFIX + "empty-slot"))
        val offHand = GuiItem(target.inventory.itemInOffHand)
        val mainHand = GuiItem(target.inventory.itemInMainHand)
        setItem(if (mainHand.type != Material.AIR) mainHand else slotEmpty, 29)
        setItem(if (offHand.type != Material.AIR) offHand else slotEmpty, 20)
        setItem(if (target.inventory.helmet != null) GuiItem(target.inventory.helmet!!) else slotEmpty, 10)
        setItem(if (target.inventory.chestplate != null) GuiItem(target.inventory.chestplate!!) else slotEmpty, 19)
        setItem(if (target.inventory.leggings != null) GuiItem(target.inventory.leggings!!) else slotEmpty, 28)
        setItem(if (target.inventory.boots != null) GuiItem(target.inventory.boots!!) else slotEmpty, 37)

        // Player Connection Information
        connectionInfoStack = GuiItem(Material.STRUCTURE_VOID)
            .setDisplayName(messageSupplier.getMessage(PREFIX + "connection.title"))
            .setLore(
                messageSupplier.getFormatted(PREFIX + "connection.ip", msgFormatter),
                messageSupplier.getFormatted(PREFIX + "connection.port", msgFormatter),
                messageSupplier.getFormatted(PREFIX + "connection.ping", msgFormatter)
            )
        setItem(connectionInfoStack, 21, player, "activecraft.connection.info")

        //player Stats
        val formatter = NumberFormat.getInstance()
        formatter.maximumFractionDigits = 2
        gameStats = GuiItem(Material.GRASS_BLOCK)
            .setDisplayName(messageSupplier.getMessage(PREFIX + "player.title"))
            .setLore(
                messageSupplier.getFormatted(PREFIX + "player.health", msgFormatter),
                messageSupplier.getFormatted(PREFIX + "player.food", msgFormatter),
                messageSupplier.getFormatted(PREFIX + "player.exp", msgFormatter),
                messageSupplier.getFormatted(PREFIX + "player.gamemode", msgFormatter)
            )
        setItem(gameStats, 12, player, "activecraft.stats.info")

        // Player Violations Information
        violationInfoStack = GuiItem(Material.COMMAND_BLOCK)
            .setDisplayName(messageSupplier.getMessage(PREFIX + "violations.title"))
            .setLore(
                messageSupplier.getFormatted(PREFIX + "violations.bans", msgFormatter),
                messageSupplier.getFormatted(PREFIX + "violations.ip-bans", msgFormatter),
                messageSupplier.getFormatted(PREFIX + "violations.warns", msgFormatter),
                messageSupplier.getFormatted(PREFIX + "violations.mutes", msgFormatter)
            )
        setItem(violationInfoStack, 30, player, "activecraft.violations.info")

        // active effects
        activeEffectsBuilder = ItemBuilder(Material.POTION)
            .displayname(messageSupplier.getMessage(PREFIX + "active-effects"))
        target.activePotionEffects.forEach { effect: PotionEffect ->
            activeEffectsBuilder.lore(
                messageSupplier.getMessage(
                    "effects." + effect.type.name.lowercase().replace("_", "-")
                )
                        + colorScheme.secondary + "; "
                        + colorScheme.secondaryAccent + effect.amplifier + colorScheme.secondary + "; "
                        + colorScheme.secondaryAccent + shortInteger(effect.duration / 20)
            )
        }

        activeEffectsStack = GuiItem(activeEffectsBuilder.build())
        val potionMeta = activeEffectsStack!!.itemMeta as PotionMeta
        potionMeta.addItemFlags(ItemFlag.HIDE_POTION_EFFECTS)
        activeEffectsStack!!.setItemMeta(potionMeta)
        // implement effectgui
        activeEffectsStack!!.addClickListener { guiClickEvent: GuiClickEvent? ->
            push(
                player,
                EffectGui(player, target).potionEffectGui.build()
            )
        }
        setItem(activeEffectsStack, 14, player, "activecraft.activeeffects")

        // Player Location Information
        // TODO: 22.08.2022 updaten
        playerLocationStack = GuiItem(Material.REDSTONE_TORCH)
            .setDisplayName(messageSupplier.getMessage(PREFIX + "player-location"))
            .setLore(
                colorScheme.secondaryAccent.toString() + target.world.name + colorScheme.secondary + "; "
                        + colorScheme.secondaryAccent + target.location.blockX + colorScheme.secondary + ", "
                        + colorScheme.secondaryAccent + target.location.blockY + colorScheme.secondary + ", "
                        + colorScheme.secondaryAccent + target.location.blockZ
            )
        setItem(playerLocationStack, 16, player, "activecraft.location")

        //Player Playtime
        playtimeStack = GuiItem(Material.CLOCK)
            .setDisplayName(messageSupplier.getMessage(PREFIX + "playtime"))
            .setLore(messageSupplier.getFormatted(PREFIX + "playtime-lore", msgFormatter))
        setItem(playtimeStack, 15, player, "activecraft.playtime")


        // Other Menus
        // Player Punishment
        violationStack = GuiItem(Material.COMMAND_BLOCK)
            .setDisplayName(messageSupplier.getMessage(PREFIX + "violations-gui"))
            .addClickListener { guiClickEvent: GuiClickEvent? ->
                push(
                    player,
                    profileMenu.violationsProfile.build()
                )
            }
        setItem(violationStack, 42)

        // Player Gamemode Switcher
        gamemodeSwitcherStack = GuiItem(Material.GRASS_BLOCK)
            .setDisplayName(messageSupplier.getMessage(PREFIX + "gamemode-switcher-gui"))
            .addClickListener { guiClickEvent: GuiClickEvent? ->
                push(
                    player,
                    profileMenu.gamemodeSwitcherProfile.build()
                )
            }
        setItem(gamemodeSwitcherStack, 32)

        // Player Storage Menu
        storageMenuStack = GuiItem(Material.CHEST)
            .setDisplayName(messageSupplier.getMessage(PREFIX + "storage-gui"))
            .addClickListener { push(player, profileMenu.storageProfile.build()) }
        setItem(storageMenuStack, 33)

        // Player Actions Menu
        actionMenuStack = GuiItem(Material.LEVER)
            .setDisplayName(messageSupplier.getMessage(PREFIX + "action-gui"))
            .addClickListener { push(player, profileMenu.actionProfile.build()) }
        setItem(actionMenuStack, 34)
    }

    companion object {
        private const val PREFIX = "profile.mainprofile."
    }
}